### ISSUE TITLE
add exec command to exec or run a command with a specific binary version

### ISSUE THREAD

**Issue by muuvmuuv** [dsherret/bvm#1]
Would be pretty useful in the future to exec or run a command with a specific binary version. For example when I want to test an application with a newer nodejs version before locking the version to it.

Syntax could be:

```shell
bvm exec node@latest ./src/index.js
bvm exec node@lts npm@^13 run build
```

Chaining binaries may be hard to parse but since a version needs to be defined it should be easy to identify which is a binary through `bin@version` syntax.

Additionally, a `--isolate` flag could be added which removes all binaries from path requiring be explicit with all binary versions for the executed command.

**Comment by dsherret** [comment]
@muuvmuuv yes, [I've wanted to add this](https://github.com/dsherret/bvm#future-improvements) and now it's easier to do than before. I like the syntax you're proposing for executing specific versions as it's familiar in npm.

Most likely this would be a little verbose:

```
bvm exec [owner-name/binary-name@version] [command-name] [args]
bvm exec [owner-name/binary-name] [command-name] [args]
bvm exec [binary-name@version] [command-name] [args]
bvm exec [binary-name] [command-name] [args]

# Examples
bvm exec nodejs/node@14.9.0 node -v
bvm exec node@14.9.0 node -v # or excluding owner when no other binaries are installed with that name
bvm exec node@14.9.0 npm -v
bvm exec node npm -v # or without owner or version for some reason
bvm exec node node -v
```

I need a better word than "binary-name"... don't want to go with "package-name", but basically node has multiple commands defined (ex. `node`, `npm`, `npx`). Binary name might be ok actually...

I'm not sure about the utility of the `node@lts npm@^13` example though. If someone wanted to use a specific version of npm, then they could install and add an alias for it via npm then execute like:

```
bvm exec node@14.9.0 npm install -g npm_6_14_8@npm:npm@6.14.8
bvm exec node@14.9.0 npm_6_14_8 -v
```

Do you have an example of where else it might be used? I think though that parsing the version out beyond the third parameter probably shouldn't be done as other binaries use that syntax (such as npm above).

I've made note of the `--isolate` flag 🙂

**Comment by muuvmuuv** [comment]
Oh yes, I haven't seen the `exec` list item. Another word would be **executable(s)** or **binaries**. 

With `^13` you can say that exec on version 13.x.x. Without knowing the exact minor and patch version number. Just aliasing `npm` wouldn't be useful because most times you execute an app through node which cannot be aliased. But because bvm is not only node and npm I guess for rust users this is also usefull?

**Comment by dsherret** [comment]
This is implemented now and looks like:

```ts
bvm exec node ~14.2
bvm exec node 12 test.js
bvm exec node 14.9.0 npm -v
````

Ended up using a space because that's what's done with other commands, slightly easier to type, and easy to work with internally.

> With ^13 you can say that exec on version 13.x.x. Without knowing the exact minor and patch version number. Just aliasing npm wouldn't be useful because most times you execute an app through node which cannot be aliased. But because bvm is not only node and npm I guess for rust users this is also usefull?

I see what you mean, but in the case of which version of npm is installed, that's handled all handled via `npm install -g npm`. I don't think there would be an easy way to integrate that into bvm without creating distributions of each npm version for bvm separately. Then I think that could work. It would probably look like `bvm exec node 14 $(bvm resolve npm ^13) run build`. I've opened https://github.com/bvm/bvm/issues/24

**Comment by muuvmuuv** [comment]
Good idea! I will test it today. Looking forward to use bvm in favor of nvm. I have my personal issues with nvm since it terribly slows down the terminal when sourcing the command...

**Comment by muuvmuuv** [comment]
Works great @dsherret !! 

```
~/Downloads took 571ms
at [$time]($style) ❯ node --version
v14.12.0

~/Downloads
at [$time]($style) ❯ npm --version
6.14.8

~/Downloads took 280ms
at [$time]($style) ❯ bvm exec node 12 -v
v12.18.4

~/Downloads took 1s250ms
at [$time]($style) ❯ bvm exec node 12 npm -v
6.14.6
```

I am on latest macOS on a 2018 MacBook Pro. Thanks :)

**Comment by dsherret** [comment]
Sweet! By the way, I realized last night that `bvm exec node 12 npm install -g <something>` will not exactly work correctly, but I'll fix that soon.

Yeah, I also tried out `nvm` and found it slow.

**Comment by dsherret** [comment]
@muuvmuuv just a heads up that I completely changed the way a lot of stuff works and it should be completely seamless on linux and mac. If you upgrade you should first delete your `~/.bvm` directory, remove any changes made to .zshrc profile for bvm, and run `npm config rm profile`.

Sorry about the instability, but yeah, it's not even 0.1 yet. I doubt there will be as big of a change in the future as this was.

**Comment by muuvmuuv** [comment]
No problem! It's still a work in progress. Now I get "No such file or directory (os error 2)" following the steps you described and running the installation sh script afterwards. Any idea? It seems it comes from `bvm-bin hidden unix-install`.

**Comment by dsherret** [comment]
@muuvmuuv thanks for letting me know about that! I didn't know removing a directory in rust would throw when it didn't exist. It worked fine on my machine because it did exist. This is fixed in 0.0.38 and I opened #41 (to add tests for the install scripts). Also, I reduced the binary size a decent amount.
